languages/java/oso/src/main/java/com/osohq/oso/Oso.java

```diff
@@ -91,7 +105,173 @@ public boolean isAllowed(Object actor, Object action, Object resource)
+   * on the resource. If they cannot read the resource, then a `NotFound` error is raised.
+   * Otherwise, a `ForbiddenError` is raised.
+   *
+   * @param actor The actor performing the request.
+   * @param action The action the actor is attempting to perform.
+   * @param resource The resource being accessed.
+   * @param checkRead If set to `false`, a `ForbiddenError` is always thrown on authorization
+   *     failures, regardless of whether the actor can read the resource. Default is `true`.
+   * @throws Exceptions.OsoException
+   */
+  public void authorize(Object actor, Object action, Object resource, boolean checkRead)
+      throws Exceptions.OsoException {
+    boolean authorized = queryRuleOnce("allow", actor, action, resource);
+    if (authorized) {
+      return;
+    }
+
+    // Authorization failure. Determine whether to throw a NotFoundException or
+    // a ForbiddenException.
+    boolean isNotFound = false;
+    if (action == readAction) {
+      isNotFound = true;
+    } else if (checkRead) {
+      boolean canRead = queryRuleOnce("allow", actor, readAction, resource);
+      if (!canRead) {
+        isNotFound = true;
+      }
+    }
+    throw isNotFound ? new Exceptions.NotFoundException() : new Exceptions.ForbiddenException();
+  }
+
+  public void authorize(Object actor, Object action, Object resource)
+      throws Exceptions.OsoException {
+    authorize(actor, action, resource, true);
+  }
+
+  /**
+   * Ensure that `actor` is allowed to send `request` to the server.
+   *
+   * <p>Checks the `allow_request` rule of a policy.
+   *
+   * <p>If the request is permitted with an `allow_request` rule in the policy, then this method
+   * returns nothing. Otherwise, this method raises a `ForbiddenError`.
+   *
+   * @param actor The actor performing the request.
+   * @param request An object representing the request that was sent by the actor.
+   * @throws Exceptions.OsoException
+   */
+  public void authorizeRequest(Object actor, Object request) throws Exceptions.OsoException {
+    boolean authorized = queryRuleOnce("allow_request", actor, request);
+    if (!authorized) {
+      throw new Exceptions.ForbiddenException();
+    }
+  }
+
+  /**
+   * Ensure that `actor` is allowed to perform `action` on a given `resource`'s `field`.
+   *
+   * <p>If the action is permitted by an `allow_field` rule in the policy, then this method returns
+   * nothing. If the action is not permitted by the policy, this method will raise a
+   * `ForbiddenError`.
+   *
+   * @param actor The actor performing the request.
+   * @param action The action the actor is attempting to perform on the field.
+   * @param resource The resource being accessed.
+   * @param field The name of the field being accessed.
+   * @throws Exceptions.OsoException
+   */
+  public void authorizeField(Object actor, Object action, Object resource, Object field)
```

Comment:
Yeah that's a good question -- I don't think there's any reason that I'd expect a `NotFoundError` if I cannot read a particular field of an object. To me a `ForbiddenError` here means, the user knows the object exists and is otherwise authorized to see it, but cannot read this particular field.

I'm not 100% sure about this though. @samscott89 had mentioned maybe always performing resource-level enforcement during this method (calling `authorize(actor, action, resource)` right at the beginning), so that could be the way that we add this eventually. But if we did that right now, it could have big performance implications and might make it a bit harder for users to understand.

Reply:
Okay makes sense. So it's on the user to only call this method on resources the user should have access to at the resource level.